server/src/main/java/org/candlepin/audit/QpidConnection.java

```diff
@@ -0,0 +1,238 @@
+     * event exchange under specific routing key. The qpid jms client allows us to
+     * do this using TopicPublisher
+     */
+    private Map<Target, Map<Type, TopicPublisher>> producerMap;
+    private static Logger log = LoggerFactory.getLogger(QpidConnection.class);
+    private InitialContext ctx = null;
+    private STATUS connectionStatus = STATUS.DOWN;
+    private QpidConfigBuilder config;
+
+    /**
+     * This class is a singleton, just in case that multiple threads
+     * try to reconnect concurrently, we want to shield ourselves
+     */
+    private static Object connectionLock = new Object();
+
+    /**
+     * Status of the connection as Candlepin sees it
+     * @author fnguyen
+     *
+     */
+    public enum STATUS {
+        CONNECTED,
+        DOWN
+    }
+
+    @Inject
+    public QpidConnection(QpidConfigBuilder config) {
+        try {
+            this.config = config;
+            ctx = new InitialContext(config.buildConfigurationProperties());
+            connectionFactory = createConnectionFactory();
+        }
+        catch (NamingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void sendTextMessage(Target target, Type type, String msg) {
+        if (connectionStatus == STATUS.DOWN) {
+            throw new RuntimeException("The Qpid connection is down, " +
+                "please use connect() to reestablish connection.");
+        }
+
+        try {
+            Map<Type, TopicPublisher> m = this.producerMap.get(target);
+            if (m != null) {
+                TopicPublisher tp = m.get(type);
+                tp.send(session.createTextMessage(msg));
+            }
+        }
+        catch (JMSException ex) {
+            log.error("Unable to send event: " + msg, ex);
+            connectionStatus = STATUS.DOWN;
+            throw new RuntimeException("Error sending event to message bus", ex);
+        }
+    }
+
+    /**
+     * This indempotent method will establish connection to qpid broker. Per JMS standard
+     * it must recreate all JMS objects such as Connection, TopicSession, TopicPublisher.
+     * @throws Exception errors during connecting to the Broker
+     */
+    public void connect() throws Exception {
+        synchronized (connectionLock) {
+            connection = newConnection();
+            log.debug("creating topic session");
+            session = createTopicSession();
+            log.info("AMQP session created successfully...");
+            Map<Target, Map<Type, TopicPublisher>> pm = Util.newMap();
+            buildAllTopicPublishers(pm);
+            producerMap = pm;
+            connectionStatus = STATUS.CONNECTED;
+        }
+
+    }
+
+    public Connection newConnection() throws JMSException {
+        log.debug("creating connection");
+        Connection conn = null;
+        conn = connectionFactory.createConnection();
+        conn.start();
+        return conn;
```

Comment:
I don't think this is a good idea. I would rather have false positive (connectionStatus = CONNECTED) then having DOWN incorrectly.

This is kind of tricky to get right, multiple threads might try to reconnect and I would rather set connectionStatus = DOWN only when I am absolutely sure the send failed.